### PR TITLE
feat(expression): Implement IS [NOT] DISTINCT FROM in backend

### LINKED ISSUE #2513: Implement IS [NOT] DISTINCT FROM
Implement `IS [NOT] DISTINCT FROM`

Reference: https://www.postgresql.org/docs/current/functions-comparison.html

Currently being tracked in #112

### LINKED ISSUE #2684: Implement IS (NOT) DISTINCT FROM` functions
Same as #2513 but this is explicitly tracked in #112

### PR DESCRIPTION
**Pull request by Graphcalibur**
## What's changed and what's your intention?

- Move implementation of `IS [NOT] DISTINCT FROM` from frontend to backend to avoid repeated evaluation of expressions (see #2582)
    - For `IS NOT DISTINCT FROM`, the frontend just binds a `NOT` to the bound `IS DISTINCT FROM` expression
- Add unit and integration tests for `IS [NOT] DISTINCT FROM`

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
Fixes #2684

### CODE REVIEW

**Review by lmatz — APPROVED** [review]
LGTM!

File: src/common/src/types/mod.rs [hunk 1]
```diff
@@ -640,6 +640,58 @@ impl ScalarRefImpl<'_> {
     }
 }
 
+/// `for_all_cmp_types` helps in matching and casting types when building comparison expressions
+///  such as <= or IS DISTINCT FROM.
+#[macro_export]
+macro_rules! for_all_cmp_variants {
```
Comment (xiangjinwu) [inline src/common/src/types/mod.rs:646]: Can we keep this macro in `crate expr` rather than in `crate common`?

**Review by xiangjinwu — APPROVED** [review]
lgtm

Some comments out of the scope of this PR:
Historically, when comparing `decimal` against `f32`, we though they should `TryInto` `decimal` or `f64`. But in PostgreSQL the common type for them is `f32`.

This affects existing comparisons including eq, ne, lt, etc as well. The macro is using `f64` and some unit tests are using `decimal`. This can be fixed in a separate PR.

File: src/common/src/types/mod.rs (hunk 1 above)
Comment (Graphcalibur) [reply src/common/src/types/mod.rs:646]: Sure. Would `/expr/src/expr/template.rs` be a suitable place for it?

**Review by xiangjinwu — APPROVED** [review]

**Pull request merged**